### PR TITLE
gh-124165: Optimize keep_top_bit() in heapq module by using _Py_bit_length()

### PR DESCRIPTION
**Pull request by visitorckw**
Optimize the keep_top_bit() function to use _Py_bit_length() instead of manually shifting bits. This allows for more efficient execution on hardware that supports specialized instructions such as bsr on x86. Additionally, the function is now marked as inline, as it is small and only used in one location, improving the potential for inlining by the compiler.

* Issue: gh-124165

### CODE REVIEW

File: Modules/_heapqmodule.c [hunk 1]
```diff
@@ -279,16 +280,10 @@ _heapq_heappushpop_impl(PyObject *module, PyObject *heap, PyObject *item)
...
 keep_top_bit(Py_ssize_t n)
 {
-    int i = 0;
-
-    while (n > 1) {
-        n >>= 1;
-        i++;
-    }
-    return n << i;
+    return (Py_ssize_t)1 << (_Py_bit_length((unsigned long)n) - 1);
```
Comment (vstinner) [inline Modules/_heapqmodule.c:288]: `_Py_bit_length(unsigned long x)` takes an unsigned long, whereas you have a `Py_ssize_t`. On Windows, `long` is only 32-bit, whereas size_t is 64-bit. Your cast is unsafe.

If you want to use _Py_bit_length(), you should add a 64-bit variant using `__builtin_clzll()` and `_BitScanReverse64()`.

File: Include/internal/pycore_bitutils.h [hunk 2]
```diff
@@ -180,6 +180,44 @@ _Py_bit_length(unsigned long x)
...
+        return (int)sizeof(uint64_t) * 8 - __builtin_clzll(x);
+    }
+    else {
+        return 0;
+    }
+#elif defined(_MSC_VER)
+    // _BitScanReverse64() is documented to search 64 bits.
+    Py_BUILD_ASSERT(sizeof(uint64_t) <= 8);
+    unsigned long msb;
+    if (_BitScanReverse64(&msb, x)) {
```
Comment (vstinner) [inline Include/internal/pycore_bitutils.h:198]: There are many warnings:
'_BitScanReverse64' undefined; assuming extern returning int [D:\a\cpython\cpython\PCbuild\_freeze_module.vcxproj]

File: Include/internal/pycore_bitutils.h [hunk 3]
```diff
@@ -180,6 +180,44 @@ _Py_bit_length(unsigned long x)
...
+#if (defined(__clang__) || defined(__GNUC__))
+    if (x != 0) {
+        // __builtin_clzll() is available since GCC 3.4.
+        // Undefined behavior for x == 0.
+        return (int)sizeof(uint64_t) * 8 - __builtin_clzll(x);
+    }
+    else {
+        return 0;
+    }
+#elif defined(_MSC_VER)
```
Comment (vstinner) [inline Include/internal/pycore_bitutils.h]: _BitScanReverse64() is not available on 32-bit Windows (x86, ARM). You should check if you're running 64-bit Windows.

File: Include/internal/pycore_bitutils.h [hunk 4]
```diff
@@ -180,6 +180,44 @@ _Py_bit_length(unsigned long x)
...
+        // __builtin_clzll() is available since GCC 3.4.
+        // Undefined behavior for x == 0.
+        return (int)sizeof(uint64_t) * 8 - __builtin_clzll(x);
+    }
+    else {
+        return 0;
+    }
+#elif defined(_MSC_VER) && defined(_WIN64)
+    // _BitScanReverse64() is documented to search 64 bits.
+    Py_BUILD_ASSERT(sizeof(uint64_t) <= 8);
```
Comment (vstinner) [inline Include/internal/pycore_bitutils.h]: You can remove this assertion, I don't see its purpose.

**Comment by visitorckw** [comment]
I noticed the warnings in the CI tests on Ubuntu related to unsigned/signed integer conversions. However, it seems most of them are not caused by my code changes. In the keep_top_bit function, I did cast the variable n, which is of type Py_ssize_t, to uint64_t. But since n will never be a negative integer, I believe we are safe.

File: Include/internal/pycore_bitutils.h [hunk 5]
```diff
@@ -180,6 +180,44 @@ _Py_bit_length(unsigned long x)
...
+    };
+    int msb = 0;
+    while (x >= 32) {
+        msb += 6;
+        x >>= 6;
+    }
+    msb += BIT_LENGTH_TABLE[x];
+    return msb;
+#endif
+}
```
Comment (vstinner) [inline Include/internal/pycore_bitutils.h:217]: Can you please add tests for this function in Modules/_testinternalcapi.c? Copy check_bit_length() and test_bit_length() to make 64-bit variant and add your test_bit_length64() to module_functions.

File: Modules/_heapqmodule.c [hunk 6]
```diff
@@ -279,16 +280,10 @@ _heapq_heappushpop_impl(PyObject *module, PyObject *heap, PyObject *item)
     return returnitem;
 }
 
-static Py_ssize_t
+static inline Py_ssize_t
 keep_top_bit(Py_ssize_t n)
 {
```
Comment (vstinner) [inline Modules/_heapqmodule.c:285]: I suggest to add:

```
Py_BUILD_ASSERT(sizeof(Py_ssize_t) <= sizeof(uint64_t));
```

File: Modules/_heapqmodule.c [hunk 7]
```diff
@@ -279,16 +280,10 @@ _heapq_heappushpop_impl(PyObject *module, PyObject *heap, PyObject *item)
...
-    int i = 0;
-
-    while (n > 1) {
-        n >>= 1;
-        i++;
-    }
-    return n << i;
+    return (Py_ssize_t)1 << (_Py_bit_length64((uint64_t)n) - 1);
```
Comment (vstinner) [inline Modules/_heapqmodule.c:290]: You may use _Py_bit_length() on 32-bit platform: when SIZEOF_SIZE_T is 4.

**Comment by rhettinger** [comment]
Thanks for the suggestion; however, the `heapq` module doesn't need `keep_top_bit()` to be optimized. It is called only once, outside of loop and only when the number of iterations exceeds 2500.  You would never be able to detect this edit in a timing.  

Also the code is already very fast, so there isn't much room for improvement with an intrinsic.  I prefer to keep it as-is because it is easier for me to maintain.

**Comment by visitorckw** [comment]
Sad to hear that.
Sorry for creating unnecessary noise, and thanks for your feedback!

**Comment by rhettinger** [comment]
It wasn't "unnecessary noise". Your suggestions are always welcome.  

This particular case wasn't wrong. It was just too microscopic to be worth having an external code path.

**Pull request closed without merging**